libs/tlv/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+{
+    let mut cursor = std::io::Cursor::new(buffer);
+    let (value, read) = D::decode(&mut cursor)?;
+
+    if read != buffer.len() {
+        Err(std::io::ErrorKind::InvalidData.into())
+    } else {
+        Ok(value)
+    }
+}
+
+/// Decodes a tagged value from a slice.
+///
+/// # Errors
+///
+/// Returns an error if the value cannot be decoded.
+///
+/// # Examples
+///
+/// ```
+/// # use tlv::{Decode, from_slice_tagged, Tag};
+/// #
+/// # fn main() -> std::io::Result<()> {
+/// let buffer = [0xff, 0x02, 0xbe, 0xef];
+/// let value: u16 = from_slice_tagged(Tag::U8(0xff), &buffer)?;
+/// assert_eq!(value, 0xbeef);
+/// # Ok(())
+/// # }
+/// ```
+pub fn from_slice_tagged<D>(tag: Tag, buffer: &[u8]) -> std::io::Result<D>
+where
+    D: Decode,
+{
+    let mut cursor = std::io::Cursor::new(buffer);
+    let (value, read) = value::decode_tagged(tag, &mut cursor)?;
+
+    if read != buffer.len() {
+        Err(std::io::ErrorKind::InvalidData.into())
+    } else {
+        Ok(value)
+    }
+}
+
+/// Decodes a tagged value from a reader, returning the value and the number of
+/// bytes read.
+///
+/// # Errors
+///
+/// Returns an error if the value cannot be decoded.
+///
+/// # Examples
+///
+/// ```
+/// # use tlv::{Decode, from_reader_tagged, Tag};
+/// #
+/// # fn main() -> std::io::Result<()> {
+/// let buffer = [0xff, 0x02, 0xbe, 0xef];
+/// let mut cursor = std::io::Cursor::new(&buffer);
+/// let (value, read) = from_reader_tagged::<u16, _>(Tag::U8(0xff), &mut cursor)?;
+/// assert_eq!(value, 0xbeef);
+/// assert_eq!(read, buffer.len());
+/// # Ok(())
+/// # }
+/// ```
+pub fn from_reader_tagged<D, R>(tag: Tag, reader: &mut R) -> std::io::Result<(D, usize)>
+where
+    D: Decode,
+    R: std::io::Read,
+{
+    value::decode_tagged(tag, reader)
+}
+
+#[cfg(test)]
+mod tests {
+    use proptest::proptest;
+
+    use crate::{
+        from_reader_tagged, from_slice, from_slice_tagged, tag::Tag, to_vec, to_vec_tagged, Decode,
+        Encode, Length,
+    };
+
+    #[test]
+    fn test_full_tlv_tag_mismatch() {
+        let value: u16 = 0xbeef;
+        let encoded = to_vec_tagged(Tag::U16(0x0102), &value).unwrap();
+
+        from_slice_tagged::<Vec<u8>>(Tag::U16(0x0103), &encoded).unwrap_err();
+    }
+
+    #[test]
+    fn test_multiple_decode() {
+        let data: [u8; 9] = [0x01, 0x02, 0xbe, 0xef, 0x99, 0x00, 0xff, 0x01, 0xfe];
+        let mut cursor = std::io::Cursor::new(&data);
+
+        let (beef, read) = from_reader_tagged::<u16, _>(Tag::U8(0x01), &mut cursor).unwrap();
+        assert_eq!(beef, 0xbeef);
+        assert_eq!(read, 4);
+
+        let (_, read) = from_reader_tagged::<(), _>(Tag::U8(0x99), &mut cursor).unwrap();
+        assert_eq!(read, 2);
+
+        let (fe, read) = from_reader_tagged::<u8, _>(Tag::U8(0xff), &mut cursor).unwrap();
+        assert_eq!(fe, 0xfe);
+        assert_eq!(read, 3);
+
+        assert_eq!(cursor.position(), data.len() as u64);
+    }
+
+    #[test]
+    fn test_encode_codegen() {
```

Comment:
This and the decode test below aren't strictly necessary, but were useful when I was figuring out how to implement `tlv-derive`.